### ISSUE TITLE
Support Audio/Video Chat (Jingle)

### ISSUE THREAD

**Issue by DanScharon** [dino/dino#154]
I know, this is a huge task, but maybe you are willing to tackle audio/video chat some day. That would be great ❤️

**Comment by edhelas** [comment]
Looking forward this one as well, I'd recommend to implement Jingle using the WebRTC stack. Movim is already doing it and I'd be please to do compatibility tests between the two clients :)

**Comment by Echolon** [comment]
Isn't it possible to include code from eg. Jitsi?

**Comment by mar-v-in** [comment]
This is on the wishlist, to be realized using libnice and gstreamer, we thus will be able to use/share some code with farstream/telepathy.

libwebrtc is horrible (it was designed to be useful for chrome and nothing else) and would require to write us large wrapping code because it is written in c++, which is hard to link against from non-c++ - including Vala.

@edhelas I guess you use xep 320+343 for Movim? While they are definitely interesting, I won't give them any priority for now...

**Comment by DanScharon** [comment]
looks like gstreamer got some [WebRTC support](http://blog.nirbheek.in/2018/02/gstreamer-webrtc.html)

**Comment by DanScharon** [comment]
see also: https://gstreamer.freedesktop.org/releases/1.14/

**Comment by WPFilmmaker** [comment]
@mar-v-in may I suggest to retry again this year? Maybe adding a page on the website specifying you are looking for people? I wrote a similar request for another project on github (which struggled to find people) and not 1 but 2 people signalled their interest to participate.

**Comment by MrSorcus** [comment]
https://www.process-one.net/blog/ejabberd-20-04/
https://gist.github.com/iNPUTmice/a28c438d9bbf3f4a3d4c663ffaa224d9#notes-for-server-admins
`ejabberd` and `conversations` implemented XEP-0215.
Will it be in the Dino in a near future?

**Comment by mar-v-in** [comment]
@MrSorcus XEP-0215 is not essential to support Audio/Video and XEP-0215 can be supported even without A/V. It is very likely that we'll implement XEP-0215 at some point, but it's only related to A/V to some degree and implementing it will not significantly help A/V implementation.

**Comment by licaon-kter** [comment]
So NATs won't be an issue and TURN won't be needed?

**Comment by MrSorcus** [comment]
@iNPUTmice said `It says most calls will require STUN or TURN. Not all. Calls within the same WiFi network or any call where you have globally routed addresses (this isn’t depended on IPv6) won’t need it.`

**Comment by licaon-kter** [comment]
@MrSorcus I'm in the known, yes. ;)

**Comment by mar-v-in** [comment]
First of all, Conversations and Dino have different target networks. Conversations mostly runs in mobile networks while Dino mostly runs on landline networks. Those use significantly different technology stacks, so what is true or relevant for mobile networks might not apply to landline networks and vice-versa.

IPv6 (typically) doesn't do NAT and thus doesn't need STUN or TURN.

Mobile networks often use symmetric NATs for IPv4. If one end of a p2p-connection has a symmetric NAT and the other one has a restricted cone NAT or a symmetric NAT, STUN cannot be used and thus TURN is required.

Landline networks typically assign a full IP address to the users router, which then typically does port-restricted cone NAT. Additionally some home routers support UPnP / NAT-PMP to forward ports (which is essentially a full cone NAT). Two restricted cone NATs can work with each other using STUN, if full cone NAT is used (or available through port forwarding), STUN can even be used with symmetric NAT on the other end. Thus TURN is much less required on landline networks, whereas UPnP / NAT-PMP are far less useful on mobile networks.

STUN is very lightweight to the server, thus public STUN servers exist (just as public DNS for example). Even non-public STUN servers are typically not password protected. In contrast TURN servers are typically password protected, because they produce a lot of traffic for the server operator.

XEP-0215 is a mechanism to find STUN and TURN servers. Another one is DNS, but DNS can't be used for credentials thus is not suited for use with TURN.

So to summarize: XEP-0215 is required for TURN. TURN should only be used if one of the call participant does not have proper IPv6 and one of the call participants is behind a symmetric NAT or similar firewall and the others are not using a public IP directly and don't have full cone NAT or port forwarding. In all other cases TURN is not required and thus XEP-0215 is not required.

**Comment by trymeouteh** [comment]
+1 for SIP calls

**Comment by licaon-kter** [comment]
@trymeouteh this is XMPP, what da you mean SIP?

**Comment by trymeouteh** [comment]
> @trymeouteh this is XMPP, what do you mean SIP?

My bad, I meant Jingle

**Comment by fiaxh** [comment]
Dino has support for audio/video calls on master.
